sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala

```diff
@@ -59,6 +62,31 @@ abstract class PropagateEmptyRelationBase extends Rule[LogicalPlan] with CastSup
     plan.output.map{ a => Alias(cast(Literal(null), a.dataType), a.name)(a.exprId) }
 
   protected def commonApplyFunc: PartialFunction[LogicalPlan, LogicalPlan] = {
+    case p: Union if p.children.exists(isEmpty) =>
+      val newChildren = p.children.filterNot(isEmpty)
+      if (newChildren.isEmpty) {
+        empty(p)
+      } else {
+        val newPlan = if (newChildren.size > 1) Union(newChildren) else newChildren.head
+        val outputs = newPlan.output.zip(p.output)
+        // the original Union may produce different output attributes than the new one so we alias
+        // them if needed
+        if (outputs.forall { case (newAttr, oldAttr) => newAttr.exprId == oldAttr.exprId }) {
+          newPlan
+        } else {
+          val newOutput = outputs.map { case (newAttr, oldAttr) =>
+            if (newAttr.exprId == oldAttr.exprId) {
+              newAttr
+            } else {
+              val newExplicitMetadata =
+                if (oldAttr.metadata != newAttr.metadata) Some(oldAttr.metadata) else None
+              Alias(newAttr, oldAttr.name)(oldAttr.exprId, explicitMetadata = newExplicitMetadata)
```

Comment:
how do you bypass `LogicalPlanIntegrity.checkIfExprIdsAreGloballyUnique` as we introduce conflicting attr id here?

Reply:
I checked if the exprId is same, and reuse the original attr if same, create a new alias if different.
```scala
if (newAttr.exprId == oldAttr.exprId) {
  newAttr
} else {
  val newExplicitMetadata =
    if (oldAttr.metadata != newAttr.metadata) Some(oldAttr.metadata) else None
  Alias(newAttr, oldAttr.name)(oldAttr.exprId, explicitMetadata = newExplicitMetadata)
}
```

Reply:
Actually the conflicting attr id is introduced by #29053, original code is:
```scala
val outputAliases = outputs.map { case (newAttr, oldAttr) =>
val newExplicitMetadata =
  if (oldAttr.metadata != newAttr.metadata) Some(oldAttr.metadata) else None
Alias(newAttr, oldAttr.name)(oldAttr.exprId, explicitMetadata = newExplicitMetadata)
```

Reply:
hmm, why doesn't `LogicalPlanIntegrity.checkIfExprIdsAreGloballyUnique` capture this...

Reply:
Current branch we only create `Alias` if the old attr id is different with new attr id, so it won't introduce the conflicting attr id.

For the #29053, it seems there is no test can cover this case so it passed..

Reply:
I think it's easy to trigger this bug: `Union(A, B)`, A is empty and we return B instead. @ulysses-you can you help to fix this bug? We have a common solution for it: `QueryPlan.transformWithNewOutput`